### ISSUE TITLE
Code compatibility and performance questions

### ISSUE THREAD

**Issue by petlist** [asherikov/qpmad#1]
Hello Alexander,

I found your code on Github and was trying to use it inside my trajectory optimisation code (also based on Eigen). I would have perhaps a feature request and a couple of remarks/observations on the performance.
Feature request:
* At the moment, the solver accepts (const) references to MatrixXd and VectorXd dense matrices. This is quite limiting for users trying to pass statically allocated matrices or matrices with single precision scalars. Moreover, possibly unnecessary evaluation and copying will be called if one attempts to pass matrix expressions. My suggestion is to use the recommended Eigen::Ref<DenseType> construction whenever possible.

Performance questions:
* For a particular instance of convex QP (not strictly convex) QPMAD returns the 'solved' status, where as the primal solution vector consist of 'nans' exclusively. The QP is obtained by linearising a nonlinear optimal control problem. If a small regularisation is added (making it strictly), QPMAD returns numeric results. Could it be possibly an issue?

* Is it possible to warm start the problem somehow?

* What it would take to support sparse matrices?

Thanks,
Peter

**Comment by asherikov** [comment]
Hi,
1. API is sloppy indeed, I'll try to find time to fix it, but I would strongly advise against using single precision floats.
2. This algorithm is designed for problems with positive definite hessians, you have to add regularization by yourself. Lack of error reporting is an issue though.
3. Nope. You can get a small performance gain by not initializing upper triangular part of the Hessian, solver uses only the lower part.
4. It would require implementing another solver.

**Comment by petlist** [comment]
Hi Alexander,

Thanks for the quick reply! I really like your implementation and would be interested to use the solver or at least benchmark against it.

Should I open a separate issue for the API improvement? It would be also fine if the solver could provide some info on its performance: final tolerance, number of iterations etc?

Another issue is that sometimes one needs the dual solution as well as the primal. Is there any chance you could make dual variable public, or provide getters for it?

Finally, in your interface the Hessian matrix is not passed as constant. Is there a reason for that? 

Best,
Peter

**Comment by asherikov** [comment]
>    Thanks for the quick reply! I really like your implementation and would
>    be interested to use the solver or at least benchmark against it.

I would actually be interested in knowing more about the problem: structure,
size, benchmarking results, etc. Pls, drop me an email to
{my_github_account}@yandex.com if you are ready to share.


>    Should I open a separate issue for the API improvement? It would be
>    also fine if the solver could provide some info on its performance:
>    final tolerance, number of iterations etc?

No, one issue is enough. I'll see what kind of information can be extracted,
but note that the solver is optimized for performance, and I am not going to
add computations that are not necessary for obtaining the solution.


>    Another issue is that sometimes one needs the dual solution as well as
>    the primal. Is there any chance you could make dual variable public, or
>    provide getters for it?

Solver computes only a subset of dual variables -- for inequality
constraints (see README).


>    Finally, in your interface the Hessian matrix is not passed as
>    constant. Is there a reason for that?

Cholesky factorization of the Hessian is computed in-place, you can reuse it
by setting `SolverParameters::HESSIAN_CHOLESKY_FACTOR` for subsequent calls.

**Comment by asherikov** [comment]
The latest version in `master` branch addresses the discussed issues (see CHANGELOG for the details)
1. Input matrices can now be statically allocated, in fact the solver can be instantiated statically as well to avoid memory allocations.
2. Added `getNumberOfInequalityIterations()` and `getInequalityDual()` methods that can be called after successful `solve()`.
3. Solver now throws an error if CHolesky factorization fails, e.g., if the Hessian is not positive definite.
4. I've added a repo with solver benchmark -> https://github.com/asherikov/qpmad_benchmark, you are welcome to contribute QP problems to this repo -> https://github.com/asherikov/qp_collection for testing.
